src/li_chao.rs

```diff
@@ -0,0 +1,99 @@
+/// A structure for answering minimum queries on a set of linear functions. Supports two
+/// operations: inserting a linear function and querying for minimum at a given point. Unlike the
+/// simplest convex hull trick implementation, the queries can be done in any order, and we can do
+/// all the calculations using integers.
+
+/// This data structure builds a static segment tree over the interval of x-coordinates
+/// [left,right), so requires O(right-left) memory. Just like normal segment trees, this could be
+/// modified to a dynamic tree when the range is huge (it can also be made persistent!).
+/// It can also probably be done as an AlCash style iterative segment tree in the static case.
+
+pub struct LiChaoTree {
+    left: i64,
+    right: i64,
+    lines: Vec<(i64, i64)>,
+}
+
+impl LiChaoTree {
+    /// Creates a new tree, built to handle queries on the interval [left, right).
+    pub fn new(left: i64, right: i64) -> Self {
+        Self {
+            left,
+            right,
+            lines: vec![(0, i64::MAX); 4 * (right - left + 1) as usize],
+        }
+    }
+
+    /// Every node in the tree has the property that the line that minimizes its midpoint is found
+    /// either in the node or one of its ancestors.  When we visit a node, we compute the winner at
+    /// the midpoint of the node. The winner is stored in the node. The loser can still possibly
+    /// beat the winner on some segment, either to the left or to the right of the current
+    /// midpoint, so we propagate it to that segment. This sequence ensures that the invariant is
+    /// kept.
+    fn add_line_impl(&mut self, mut m: i64, mut b: i64, ix: usize, l: i64, r: i64) {
+        let x = (l + r) / 2;
```

Comment:
Hm I think you could do something like:
`let (ref mut m_ix, ref mut b_ix) = self.lines[ix];`

Reply:
Done. Would something like `let (m_ix, b_ix) = &self.lines[ix]` also work, or is there something wrong with that?

Reply:
I think you would have a typing error there because a reference to a pair is not the same as a pair of references. They have to be converted.

Reply:
Whoa never mind you're absolutely right! I'm confused though. How come
`let (m_ix, b_ix) = &mut self.lines[ix]` works, but
`let (m_ix, b_ix): (&mut i64, &mut i64) = &mut self.lines[ix]` doesn't??